src/loaders/umx_load.c

```diff
@@ -43,97 +42,374 @@ const struct format_loader libxmp_loader_umx = {
 	umx_load
 };
 
-static int umx_test(HIO_HANDLE *f, char *t, const int start)
+/* UPKG parsing partially based on Unreal Media Ripper (UMR) v0.3
+ * by Andy Ward <wardwh@swbell.net>, with additional fixes & updates
+ * by O. Sezer - see git repo at https://github.com/sezero/umr.git
+ */
+typedef int32 fci_t;		/* FCompactIndex */
+
+#define UPKG_HDR_TAG		0x9e2a83c1
+
+struct _genhist {	/* for upkg versions >= 68 */
+	int32 export_count;
+	int32 name_count;
+};
+
+struct upkg_hdr {
+	uint32 tag;	/* UPKG_HDR_TAG */
+	int32 file_version;
+	uint32 pkg_flags;
+	int32 name_count;	/* number of names in name table (>= 0) */
+	int32 name_offset;		/* offset to name table  (>= 0) */
+	int32 export_count;	/* num. exports in export table  (>= 0) */
+	int32 export_offset;		/* offset to export table (>= 0) */
+	int32 import_count;	/* num. imports in export table  (>= 0) */
+	int32 import_offset;		/* offset to import table (>= 0) */
+
+	/* number of GUIDs in heritage table (>= 1) and table's offset:
+	 * only with versions < 68. */
+	int32 heritage_count;
+	int32 heritage_offset;
+	/* with versions >= 68:  a GUID, a dword for generation count
+	 * and export_count and name_count dwords for each generation: */
+	uint32 guid[4];
+	int32 generation_count;
+#define UPKG_HDR_SIZE 64			/* 64 bytes up until here */
+	struct _genhist *gen;
+};
+/* compile time assert for upkg_hdr size */
+typedef int _check_hdrsize[2 * (offsetof(struct upkg_hdr, gen) == UPKG_HDR_SIZE) - 1];
+
+#define UMUSIC_IT	0
+#define UMUSIC_S3M	1
+#define UMUSIC_XM	2
+#define UMUSIC_MOD	3
+#define UMUSIC_WAV	4
+#define UMUSIC_MP2	5
+
+static const char *mustype[] = {
+	"IT", "S3M", "XM", "MOD",
+	"WAV", "MP2", NULL
+};
+
+/* decode an FCompactIndex.
+ * original documentation by Tim Sweeney was at
+ * http://unreal.epicgames.com/Packages.htm
+ * also see Unreal Wiki:
+ * http://wiki.beyondunreal.com/Legacy:Package_File_Format/Data_Details
+ */
```

Comment:
Not a real review yet (sorry, it is late, will do tomorrow) but I checked docs/formats/ and currently there is no UMX doc. Both of these links seem dead, but would it be worth pulling one of these from Wayback (or somewhere else)?

Reply:
Wayback links: 
https://web.archive.org/web/20080122034944/http://unreal.epicgames.com/Packages.htm
https://web.archive.org/web/20201023045227/http://wiki.beyondunreal.com/Legacy:Package_File_Format
https://web.archive.org/web/20201023062426/http://wiki.beyondunreal.com/Legacy:Package_File_Format/Data_Details

Don't know whether it's worth adding something under docs/

Reply:
OK, I copied the attached txt from beyondunreal wiki. 
Would it be appropriate to add under docs/?
[upkg.txt](https://github.com/libxmp/libxmp/files/5911580/upkg.txt)

Reply:
Well, I went ahead and added the documentation under docs/formats/ with the last commit.